### PR TITLE
[usb] Mass storage device driver.

### PR DESCRIPTION
**Pull request by MichalBlk**
Introduce a USB mass storage device driver.

### CODE REVIEW

File: launch [hunk 1]
```diff
@@ -81,6 +81,8 @@ CONFIG = {
                 'binary': 'qemu-mimiker-mipsel',
                 'options': [
                     '-device', 'VGA',
+                    # '-drive', 'if=none,id=stick,file=usb_drive.img',
```
Comment (hadarai) [inline launch:84]: Why is it commented out?
If you mean to leave it here, add one more line of comment about why is that

File: sys/drv/umass.c [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+  off_t offset = uio->uio_offset;
+  int error = 0;
+
+  /* Obtain the LBA of the first block invloved. */
+  uint32_t start = offset / block_size;
+
+  /* Obtain the LBA of the last block involved. */
+  uint32_t end = roundup(offset + uio->uio_resid, block_size) / block_size;
+
+  /* How many blocks is it?
```
Comment (hadarai) [inline sys/drv/umass.c:401]: Excuse me, but who is supposed to answer that question? 
If that question is supposed to explain the value of `nblocks` then rewrite it's a sentence

File: sys/drv/umass.c [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+   * If we won't overwrite the edge blocks, we have to preserve their content.
+   * To achive this, we have to read their current data.
+   */
+
+  scsi_rw_10_t rw10 = (scsi_rw_10_t){
+    .opcode = READ_10,
+    .length = htobe16(1),
+  };
+
+  /* Should we bother about the first block? */
```
Comment (hadarai) [inline sys/drv/umass.c:420]: same as above

File: sys/drv/umass.c [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+  if (offset % block_size) {
+    rw10.addr = htobe32(start);
+    error = umass_transfer(dev, buf, block_size, USB_DIR_INPUT, &rw10,
+                           sizeof(scsi_rw_10_t));
+    if (error)
+      goto end;
+  }
+
+  /* Should we bother about the last block? */
```
Comment (hadarai) [inline sys/drv/umass.c:429]: same as above

File: sys/drv/usb.c [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#include <dev/usb.h>
+
+usb_buf_t *usb_buf_alloc(void) {
+  /* Not implemented! */
```
Comment (hadarai) [inline sys/drv/usb.c:4]: There shouldn't be something like `assert(false)` then?

**Comment by cahirwpz** [comment]
@MichalBlk Please resolve conflicts.

**Comment by cahirwpz** [comment]
@MichalBlk Could you resolve conflicts?

**Comment by MichalBlk** [comment]
This PR is once again ready for code review!

File: sys/drv/umass.c [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+   */
+
+  /* Obtain the number and length of the logical blocks. */
+  if ((error = umass_read_capacity(dev)))
+    return ENXIO;
+
+  umass_print(dev);
+
+  /* Prepare /dev/umass interface. */
+  devfs_makedev(NULL, "umass", &umass_ops, dev, NULL);
```
Comment (cahirwpz) [inline sys/drv/umass.c:451]: We're trying to get rid of this function. Please replace it with `devfs_makedev_new`.

Comment (MichalBlk) [reply sys/drv/umass.c]: Done.

File: sys/drv/umass.c [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+    .opcode = (dir == USB_DIR_INPUT) ? READ_10 : WRITE_10,
+    .addr = htobe32(start),
+    .length = htobe16(nblocks),
+  };
+  error = umass_transfer(dev, &rw10, sizeof(scsi_rw_10_t), dir, buf, size);
+  if (error)
+    goto end;
+
+  if (dir == USB_DIR_INPUT) {
+    /* Copy the data to the user. */
```
Comment (cahirwpz) [inline sys/drv/umass.c:384]: Ditto

**Review by cahirwpz — COMMENTED** [review]
I find code to be simple and readable, but I cannot easily state it's correct. It's sometimes difficult to find correspondence between the code and the documentation. Some comments are useless.

File: sys/drv/umass.c [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+#include <dev/scsi.h>
+#include <dev/usb.h>
+#include <dev/umass.h>
+#include <sys/vnode.h>
+
+/* We assume that block size >= 512. */
+#define UMASS_MIN_BLOCK_SIZE 512
+
+typedef struct umass_state {
+  uint32_t next_tag;                    /* next tag to grant */
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/drv/umass.c]: I changed the description to indicate that the tag is related to the CBW/CSW structures.

File: sys/drv/umass.c [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+
+  if (*umass->revision)
+    klog("revision: %s", umass->revision);
+
+  klog("number of logical blocks: %u", umass->nblocks);
+  klog("block size: %u", umass->block_size);
+}
+
+/*
+ * Command Block Wrapper handling functions.
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/drv/umass.c:85]: Done! Each transfer function has been given a reference to the relevant section of the specification.

File: sys/drv/umass.c [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+  }
+
+  /* Has there been any error? */
+  if (csw->bCSWStatus & CSWSTATUS_FAILED)
+    return EIO;
+
+  return 0;
+}
+
+static int csw_receive(device_t *dev, umass_bbb_csw_t *csw, usb_buf_t *buf) {
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/drv/umass.c]: Yes. Now it does :).

File: sys/drv/umass.c [hunk 11]
```diff
@@ -0,0 +1,463 @@
...
+    return error;
+
+  return EIO;
+}
+
+/*
+ * Bulk-Only transfer functions.
+ */
+
+static int umass_transfer(device_t *dev, void *cmd, uint8_t cmdsize,
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/drv/umass.c:184]: Fixed.

File: sys/drv/umass.c [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+    .length = htobe16(SHORT_INQUIRY_LENGTH),
+  };
+  int error = 0;
+
+  if ((error = umass_transfer(dev, &inq, sizeof(scsi_inquiry_t), USB_DIR_INPUT,
+                              &inq_data, SHORT_INQUIRY_LENGTH)))
+    return error;
+
+  /* We only handle direct access block devices (i.e. command set SBC-3). */
+  if (SID_TYPE(&inq_data) != 0x00)
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/drv/umass.c]: Good point. `T_DIRECT` it is.

File: sys/drv/umass.c [hunk 13]
```diff
@@ -0,0 +1,463 @@
...
+    return ENXIO;
+
+  /* The device must be connected to the logical unit. */
+  if (SID_QUAL(&inq_data) != SID_QUAL_LU_CONNECTED)
+    return ENXIO;
+
+  /* Response format should be 0x02 (0x01 is possible but obsolete). */
+  uint8_t format = SID_FORMAT(&inq_data);
+  if (format != 0x02 && format != 0x01)
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/drv/umass.c]: As FreeBSD assumes that the format is equal to 2, the snippet has been removed.

File: sys/drv/umass.c [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+  /* Copy supplied strings. */
+  umass_state_t *umass = dev->state;
+  memcpy(umass->vendor, inq_data.vendor, SID_VENDOR_SIZE);
+  memcpy(umass->product, inq_data.product, SID_PRODUCT_SIZE);
+  memcpy(umass->revision, inq_data.revision, SID_REVISION_SIZE);
+
+  return 0;
+}
+
+/* Perform the request sense command. */
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/drv/umass.c]: Removed.

File: sys/drv/umass.c [hunk 15]
```diff
@@ -0,0 +1,463 @@
...
+    .length = sizeof(scsi_sense_data_t),
+  };
+  /* XXX: in the future, some recovery logic may be needed.
+   * FTTB, simply performing the command is sufficient. */
+  return umass_transfer(dev, &sen, sizeof(scsi_request_sense_t), USB_DIR_INPUT,
+                        &sen_data, sizeof(scsi_sense_data_t));
+}
+
+/*
+ * For the read cpapcity request to succeed, most devices will require to
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/drv/umass.c]: Fixed.

File: sys/drv/umass.c [hunk 16]
```diff
@@ -0,0 +1,463 @@
...
+  uint32_t size = nblocks * block_size;
+
+  /* NOTE: we assume that number of blocks to read is <= `UINT16_MAX`. */
+  if (nblocks > UINT16_MAX)
+    return EINVAL;
+
+  void *buf = kmalloc(M_DEV, size, M_WAITOK);
+
+  if (dir == USB_DIR_OUTPUT) {
+    /* Copy data from the user. */
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply sys/drv/umass.c]: Removed.

File: sys/drv/umass.c (hunk 7 above)
Comment (MichalBlk) [reply sys/drv/umass.c]: Removed.

**Review by cahirwpz — APPROVED** [review]
LGTM

**Pull request merged**